src/manage_passphrase.rs

```diff
@@ -0,0 +1,159 @@
+use libc::c_char;
+use sequoia_openpgp::packet::key::{
+    KeyRole, PrimaryRole, SecretKeyMaterial, SecretParts, SubordinateRole,
+};
+use sequoia_openpgp::packet::Key;
+use sequoia_openpgp::{crypto::Password, Fingerprint};
+use sequoia_openpgp::{Cert, Packet};
+use std::ffi::CStr;
+
+use crate::pep::{Error, PepIdentity, Result, Session};
+
+use crate::ErrorCode;
+
+ffi!(
+    fn pgp_manage_passphrase(
+        session: &mut Session,
+        identity: *const PepIdentity,
+        old_passphrase: *const c_char,
+        passphrase: *const c_char) -> Result<()> {
+        let fpr_str = unsafe { identity.as_ref() }
+            .map(|i| i.fingerprint())
+            .flatten()
+            .ok_or_else(|| illegal_value("no fingerprint on identity"))?
+            .to_str()
+            .map_err(|_| illegal_value("cannot convert identity fingerprint to a string"))?;
+
+        let fingerprint = Fingerprint::from_hex(fpr_str)
+            .map_err(|_| illegal_value("cannot create fingerprint from hex value"))?;
+
+        let (cert, _) = session.keystore().cert_find(fingerprint.clone(), true)?;
+
+        if !cert.is_tsk() {
+            return Err(illegal_value("have no secret key"));
+        }
+
+        let new_passphrase = unsafe { check_cstr!(passphrase) }
+            .to_str()
+            .map_err(|_| illegal_value("new passphrase cannot be converted to string"))?;
+
+        let remove_passphrase = { new_passphrase.is_empty() };
+
+        let new_passphrase = Password::from(new_passphrase);
+        let old_passphrase = unsafe { check_cstr!(old_passphrase) }
+            .to_str()
+            .map_err(|_| illegal_value("passphrase cannot be converted to string"))
+            .map(|s| Password::from(s))?;
+
+        let decrypted_packets = decrypted_packets(&cert, &old_passphrase)?;
+
+        let cert = cert
+            .insert_packets(decrypted_packets)
+            .map_err(|_| illegal_value("cannot not re-insert decrypted packets"))?;
+
+        let cert = if remove_passphrase {
+            cert
+        } else {
+            let encrypted_packets = encrypted_packets(&cert, &new_passphrase)?;
+            cert.insert_packets(encrypted_packets)
+                .map_err(|_| illegal_value("cannot not re-insert encrypted packets"))?
+        };
+
+        // The way `cert_save` handles certificate merging makes this step necessary.
+        // Otherwise, you'll end up with secrets encrypted with the old key.
+        session.keystore().cert_delete(fingerprint)?;
+
+        session.keystore().cert_save(cert)?;
+
+        Ok(())
```

Comment:
why don't you return `session.keystore().cert_save(cert)` ?

Reply:
The caller (ultimately core), has no use for the return value of `cert_save`.